src/utils/theme.js

```diff
@@ -0,0 +1,41 @@
+import React from 'react'
+import _ from 'lodash/fp'
+
+let ThemeContext = React.createContext()
+
+export let ThemeProvider = ({ children, theme }) => (
+  <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
+)
+
+export let mergeNestedTheme = (key = '', theme) =>
+  _.flow(
+    _.pickBy((val, k) => _.startsWith(`${key}.`, k)),
+    _.mapKeys(x => _.replace(`${key}.`, '', x)),
+    _.defaults(theme)
+  )(theme || React.useContext(ThemeContext))
+
+export let mergeNestedThemePath = (path, theme) =>
+  _.reduce(
+    (acc, key) => mergeNestedTheme(key, acc),
+    theme || React.useContext(ThemeContext),
+    path || []
+  )
+
+export let ThemeConsumer = ({ children, path }) =>
+  children(mergeNestedThemePath(path))
+
+export let withTheme = (defaults = {}, name) => Component => ({
+  theme: propTheme,
+  ...props
+}) => {
+  name = name || Component.displayName || Component.name
```

Comment:
I think it might be better to force explicit opt in (so that you can only override things that have passed their name into withTheme). This will do two things: 1. If there is no name, skip over all the merging stuff and just treat it like regular context, and 2. Allow us to not worry about getting component names exactly right (otherwise tweaking display names would actually become breaking changes)